File: src/com/Torvald/Terrarum/GameControl/KeyToggler.java
package com.Torvald.Terrarum.GameControl;

import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Input;

public class KeyToggler {

    private static boolean[] currentState = new boolean[256];
    private static boolean[] isPressed = new boolean[256];
    private static boolean[] isToggled = new boolean[256];

    public static boolean isOn(int key){
        return currentState[key];
    }

    public static void update(Input input){
        for (int i = 0; i < 256; i++) {
            if (input.isKeyDown(i)) {
                isPressed[i] = true;
            } else {
                isPressed[i] = false;
            }
        }

        for (int i = 0; i < 256; i++){
            if (isPressed[i] && !currentState[i] && !isToggled[i]){
                currentState[i] = true;
                isToggled[i] = true;
            }
            else if(isPressed[i] && currentState[i] && !isToggled[i]){
                currentState[i] = false;
                isToggled[i] = true;
            }

            if (!isPressed[i] && isToggled[i]){
                isToggled[i] = false;
            }
        }
    }

    public static void forceSet(int key, boolean b) {
        currentState[key] = b;
        isToggled[key] = true;
    }
	
}

File: src/com/Torvald/Terrarum/GameControl/KeyToggler.kt
package com.Torvald.Terrarum.GameControl

import org.newdawn.slick.GameContainer
import org.newdawn.slick.Input

object KeyToggler {

    private val currentState = BooleanArray(256)
    private val isPressed = BooleanArray(256)
    private val isToggled = BooleanArray(256)

    fun isOn(key: Int): Boolean {
        return currentState[key]
    }

    fun update(input: Input) {
        for (i in 0..255) {
            if (input.isKeyDown(i)) {
                isPressed[i] = true
            }
            else {
                isPressed[i] = false
            }
        }

        for (i in 0..255) {
            if (isPressed[i] && !currentState[i] && !isToggled[i]) {
                currentState[i] = true
                isToggled[i] = true
            }
            else if (isPressed[i] && currentState[i] && !isToggled[i]) {
                currentState[i] = false
                isToggled[i] = true
            }

            if (!isPressed[i] && isToggled[i]) {
                isToggled[i] = false
            }
        }
    }

    fun forceSet(key: Int, b: Boolean) {
        currentState[key] = b
        isToggled[key] = true
    }

}
